Make ImportedIdentities target protocol versions and KDFs (#18)
tlswg/draft-ietf-tls-external-psk-importer: draft-ietf-tls-external-psk-importer.md

# Overview

Key importers mirror the concept of key exporters in TLS in that they diversify a key
based on some contextual information before use in a connection. In contrast to key exporters,
wherein differentiation is done via an explicit label and context string, the key importer
defined herein uses an optional context string along with a target protocol and KDF
identifier to differentiate an external PSK into one or more PSKs for use.

Imported keys do not require negotiation for use, as a client and server will not agree upon
identities if not imported correctly. Thus, importers induce no protocol changes with
the exception of expanding the set of PSK identities sent on the wire. Endpoints may
incrementally deploy PSK importer support by offering non-imported keys for TLS versions
prior to TLS 1.3. (Negotiation and use of imported PSKs requires both endpoints support

> the importer API described herein.)

Reviewer: I don't think that you need the parenthetical here any more.

Also, you should point out that imported PSKs and non-imported PSKs are distinguished by having different identifiers.  It seems obvious, but we know how little that matters when people are looking to play dumb.

Author: That seems reasonable. The parenthetical meant to make it painfully obvious that both sides need to import to use an external PSK. Replacing it with the suggested "non-imported and imported identities are different" text does the trick, too!

Clients which import external keys TLS MUST NOT use these keys for any other purpose.
Moreover, each external PSK MUST be associated with at most one hash function.

## Terminology {#terminology}

> - External PSK (EPSK): A PSK established or provisioned out-of-band, i.e., not from a TLS

Reviewer: I think that you want a term for the PSK that you provision (this one) AND the PSK that you use with the target protocol and KDF.  You don't seem to have the latter, but I think that you are talking imported PSK (IPSK), right?

Author: Yep! I'll add an Imported PSK definition.

connection, which is a tuple of (Base Key, External Identity, Hash). The associated hash

> function may be undefined.

Reviewer: I think that you should drop the "hash is undefined" piece.  There is no reason to have that.  These are for exclusive use and provision of the key can just assume a default hash function if they want to avoid the overhead of having to say "SHA-256" every time.

Author: That seems reasonable to me. Dropped!

- Base Key: The secret value of an EPSK.
- External Identity: The identity of an EPSK.
- Imported Identity: The identity of a PSK as sent on the wire.
- Target protocol: The protocol for which a PSK is imported for use.
- Target KDF: The KDF for which a PSK is imported for use.

# Key Import

> A key importer takes as input an EPSK with external identity `external_identity` and base key `epsk`,

Reviewer: and a hash function.

Author: That's covered in the remainder of this sentence.

as defined in {{terminology}}, along with an optional context, and transforms it into a set of PSKs
and imported identities for use in a connection based on supported (target) protocols and KDFs. In particular,
for each supported target protocol `target_protocol` and KDF `target_kdf`, the importer constructs
an ImportedIdentity structure as follows:

~~~

> uint8 KDFID[2];

Reviewer: Really, why not uint16?  In my opinion, the use of an array for TLS versions was a terrible idea, don't repeat that.  People will start to think that each uint8 has semantics on its own and then you will never be able to change those bits without some sort of misinterpretation.

Author: I only went with it for consistency with CipherSuite. I'm fine with uint16. I'll make that change.

struct {
   opaque external_identity<1...2^16-1>;
   opaque context<0..2^16-1>;
   uint16 target_protocol;
   KDFID target_kdf;

> } ImportedIdentity;

Reviewer: Is this structure what would be included in the TLS handshake as a PSK ID?

Author: Yep! I'll spell that out clearly.

~~~

The list of KDFID values is maintained by IANA as described in {{IANA}}. External PSKs MUST NOT
be imported for versions of (D)TLS 1.2 or prior versions. See {{rollout}} for discussion on
how imported PSKs for TLS 1.3 and non-imported PSKs for earlier versions co-exist for incremental
deployment.

ImportedIdentity.context MUST include the context used to derive the EPSK, if any exists.
If the EPSK is a key derived from some other protocol or sequence of protocols,
ImportedIdentity.context MUST include a channel binding for the deriving protocols
{{!RFC5056}}.  If any secrets are agreed in earlier protocols they SHOULD be included

> in ImportedIdentity.context {{CCB}}.

Reviewer: This last bit is the kicker for using ImportedIdentity in a handshake.  We should include those secrets in the derivation of the base key, and we might use those secrets in the derivation of the context, but we shouldn't be transmitting them.

Author: I'm just removing that sentence (for now).

ImportedIdentity.target_protocol MUST be the (D)TLS protocol version for which the

> PSK is being imported. For example, TLS 1.3 {{!RFC8446}} and QUICv1 {{!I-D.ietf-quic-transport}}

Reviewer: Is this the on-wire version number or the abstract version number?

We identify DTLS 1.3 the same as TLS 1.3, with 0x0304, but we identity DTLS 1.2 in two ways: either 0x0303 or 0xfdfd.  I'm sure people will want to know which is used here.

```suggestion
PSK is being imported. For example, TLS 1.3 {{!RFC8446}} and QUICv1 {{!QUIC=I-D.ietf-quic-transport}}
```

Author: This is the on-wire version number. (I left out DTLS since DTLS 1.3 uses 0x0304 and I removed import support for (D)TLS 1.2.)

> MUST use 0x0304. Note that this means future versions of TLS will increase the number of PSKs

Reviewer: ```suggestion
use 0x0304. Note that this means future versions of TLS will increase the number of PSKs
```

Author: Will fix!

derived from an external PSK.

An imported PSK (IPSK) with base key 'ipskx' bound to this identity is then computed as follows:

~~~
   epskx = HKDF-Extract(0, epsk)
   ipskx = HKDF-Expand-Label(epskx, "derived psk",
                             Hash(ImportedIdentity), L)
~~~

> L is the length of the hash function associated with ImportedIdentity.target_kdf, as this

Reviewer: This suddenly assumes that the target KDF has a hash function.  You might say that this value is defined by the target KDF, and for those that are hash-based it is the length of the hash function output.

Author: Fixed!

is required for the imported PSK to be of suitable length with compatible ciphersuites.

> The hash function used for HKDF {{!RFC5869}} is that which is associated with the external PSK.

Reviewer: Paragraph break.

```suggestion

The hash function used for HKDF {{!RFC5869}} is that which is associated with the external PSK.
```

Author: 👍

It is not the hash function associated with ImportedIdentity.target_kdf. If no hash function

> is specified, SHA-256 MUST be used. Differentiating epsk by ImportedIdentity.target_kdf ensures

Reviewer: ```suggestion
is specified, SHA-256 MUST be used. Diversifying EPSK by ImportedIdentity.target_kdf ensures
```

that an imported PSK is only used as input keying material to at most one KDF, thus satisfying
the requirements in {{!RFC8446}}.

> Endpoints MUST import an ipskx for each target KDF they support. For example, importing a key for

Reviewer: ```suggestion
Endpoints generate an Imported PSK for each target KDF they support. For example, importing a key for
```

I don't think that you want a MUST here.  If I don't *want* to use this EPSK with a given KDF, I shouldn't be forced to.  For instance, if I have an EPSK that is "strong", I might not want to use it with a weaker KDF and fall back to ECDH instead.  That sounds crazy now that I write it down, but I don't think we can proscribe that behaviour.

Author: This should be for each ciphersuite. I'll drop the normative MUST and swap "for each target KDF" with "for each offered ciphersuite". (One shouldn't use a PSK that won't work with a ciphersuite offered on the wire.)

TLS_AES_128_GCM_SHA256 and TLS_AES_256_GCM_SHA384 would yield two PSKs, one for HKDF-SHA256 and
another for HKDF-SHA384. In contrast, if TLS_AES_128_GCM_SHA256 and TLS_CHACHA20_POLY1305_SHA256
are supported, only one derived key is necessary.

The resulting IPSK base key 'ipskx' is then used as the binder key in TLS 1.3 with identity
ImportedIdentity. With knowledge of the supported KDFs, one may import PSKs before the start
of a connection.

EPSKs may be imported for early data use if they are bound to protocol settings and configurations that would
otherwise be required for early data with normal (ticket-based PSK) resumption. Minimally, that means ALPN,
QUIC transport settings, etc., must be provisioned alongside these EPSKs.

# Deprecating Hash Functions

If a client or server wish to deprecate a hash function and no longer use it for TLS 1.3,

> they may remove the corresponding KDF from the set of target KDFs used for importing keys.

Reviewer: MAY?

```suggestion
they remove the corresponding KDF from the set of target KDFs used for importing keys.
```

Author: Will fix!

> This does not affect the KDF operation used to derive concrete PSKs.

Reviewer: "concrete"?

```suggestion
This does not affect the KDF operation used to derive Imported PSKs.
```

Author: Will fix!

as TLS 1.3 with importers. However, critically, the derived PSK will not be the same since
the importer differentiates the PSK via the identity, target protocol, and target KDF. Thus,
PSKs imported for TLS 1.3 are distinct from those used in TLS 1.2, and thereby avoid
cross-protocol collisions. Note that this does not preclude endpoints from using non-imported
PSKs for TLS 1.2. Indeed, this is necessary for incremental deployment.

# Security Considerations

DISCLAIMER: This is a WIP draft and has not yet seen significant security analysis.

# Privacy Considerations

External PSK identities are typically static by design so that endpoints may use them to
lookup keying material. However, for some systems and use cases, this identity may become a
persistent tracking identifier.

# IANA Considerations {#IANA}

This specification introduces a new registry for TLS KDF identifiers and defines the following
KDFID values:

+--------------------+-------------+
| Description        | Value       |
+--------------------+-------------+
| Reserved           | {0x00,0x00} |
|                    |             |
| HKDF_SHA256        | {0x00,0x01} |
|                    |             |

> | HKDF_SHA384        | {0x00,0x02} |

Reviewer: See above regarding the use of tuples to identify things.

Author: Yep, good call. I'll fix it.

+--------------------+-------------+

New KDFID values are allocated according to the following process:

- Values with the first in the range 0-254 (decimal) are assigned via Specification Required {{!RFC8126}}.

> - Values with the first byte 255 (decimal) are reserved for Private Use {{!RFC8126}}.

Reviewer: You can use ranges: 0x0000-0xfeff and 0xff00-0xffff.

Author: Thanks -- will fix!

are supported, only one derived key is necessary.

The resulting IPSK base key 'ipskx' is then used as the binder key in TLS 1.3 with identity
ImportedIdentity. With knowledge of the supported KDFs, one may import PSKs before the start
of a connection.

EPSKs may be imported for early data use if they are bound to protocol settings and configurations that would
otherwise be required for early data with normal (ticket-based PSK) resumption. Minimally, that means ALPN,
QUIC transport settings, etc., must be provisioned alongside these EPSKs.

# Deprecating Hash Functions

If a client or server wish to deprecate a hash function and no longer use it for TLS 1.3,
they remove the corresponding KDF from the set of target KDFs used for importing keys.
This does not affect the KDF operation used to derive Imported PSKs.

# Incremental Deployment {#rollout}

Recall that TLS 1.2 permits computing the TLS PRF with any hash algorithm and PSK.
Thus, an EPSK may be used with the same KDF (and underlying HMAC hash algorithm)
as TLS 1.3 with importers. However, critically, the derived PSK will not be the same since

> the importer differentiates the PSK via the identity, target protocol, and target KDF. Thus,

Reviewer 2: Does this text need to be tweaked if we believe EPSKs can't be used with TLS 1.2 at all?

Author: Hmm, I don't think so? What did you have in mind?

ImportedIdentity.context MUST include a channel binding for the deriving protocols
{{!RFC5056}}.  If any secrets are agreed in earlier protocols they SHOULD be included
in ImportedIdentity.context {{CCB}}.

ImportedIdentity.target_protocol MUST be the (D)TLS protocol version for which the
PSK is being imported. For example, TLS 1.3 {{!RFC8446}} and QUICv1 {{!I-D.ietf-quic-transport}}
MUST use 0x0304, whereas TLS 1.2 uses 0x0303 and DTLS 1.2 uses 0xFEFD. Note that this means
future versions of TLS will increase the number of PSKs derived from an external PSK.

An imported PSK (IPSK) with base key 'ipskx' bound to this identity is then computed as follows:

~~~
   epskx = HKDF-Extract(0, epsk)
   ipskx = HKDF-Expand-Label(epskx, "derived psk",
                             Hash(ImportedIdentity), L)
~~~

L is the length of the hash function associated with ImportedIdentity.target_kdf, as this
is required for the imported PSK to be of suitable length with compatible ciphersuites.
The hash function used for HKDF {{!RFC5869}} is that which is associated with the external PSK.
It is not the hash function associated with ImportedIdentity.target_kdf. If no hash function

> is specified, SHA-256 MUST be used. Differentiating epsk by ImportedIdentity.target_kdf ensures

Author: Yep! I can add a column to the table indicating that.

lookup keying material. However, for some systems and use cases, this identity may become a
persistent tracking identifier.

# IANA Considerations {#IANA}

This specification introduces a new registry for TLS KDF identifiers and defines the following
KDFID values:

+--------------------+-------------+
| Description        | Value       |
+--------------------+-------------+
| Reserved           | {0x00,0x00} |
|                    |             |
| HKDF_SHA256        | {0x00,0x01} |
|                    |             |
| HKDF_SHA384        | {0x00,0x02} |
|                    |             |

> | TLS12_PRF_MD5      | {0x00,0x03} |

Author: Nope -- we can remove it, and SHA384!

|                    |             |

> | TLS12_PRF_MD5SHA1  | {0x00,0x04} |

Author: This is me being lazy :) I can make it `TLS10_PRF_MD5SHA1`.

|                    |             |

> | TLS12_PRF_SHA256   | {0x00,0x05} |

Reviewer 2: Prior to this PR, we had exactly one more for TLS 1.2 and earlier, to reflect their existing lack of key separation. Is the thinking that we'll retroactively add in some key separation into TLS 1.2? Protocol-wise, that seems doable enough. PSKs are negotiated after and version cipher suites in TLS 1.2, the negotiation is extra straightforward. But I think we should describe this explicitly.

We also have this sentence in the introduction, which doesn't quite contradict it but may need some follow-up text:

> TLS 1.2 {{!RFC5246}}, in contrast, has no such requirement, as a PSK may be used with any hash algorithm and the TLS 1.2 PRF.

The other consideration is what we're keeping TLS 1.2 compatibility around for. These are the scenarios I can imagine:

1. This is a new application with no legacy TLS 1.2 PSK deployment at all. It can just assume TLS 1.3+.
2. This is a new application with no legacy TLS 1.2 PSK deployment, but for some odd reason it can't just assume TLS 1.3+?
3. This is an application which previously used TLS 1.2 PSKs. It wishes to smooth migrate to TLS 1.3, which means dual-1.2/1.3 endpoints must exist for a period of time.
4. Same as above, but this time the underlying TLS library wishes to enable TLS 1.3 by default, so no API calls to the TLS stack may change.

Case 1 is easy. It just doesn't care about TLS 1.2 logic in this draft, either before or after this PR.

Case 2 is odd. It can use the TLS 1.2 logic in this draft. Before this PR is definitely fine as it requires no new TLS 1.2 behavior. After this PR may be tricky, depending on exactly why you can't assume TLS 1.3. If it's because, say, you need to run on an older version of some OS without TLS 1.3, you're limited by what the older TLS 1.2 API exposed. That may or may not be sufficient for this scheme. (E.g. OpenSSL does some callback thing, so you might be able to implement it?? Not sure. But declaratively configuring a set of PSKs won't know to match the KDF.) If you need TLS 1.2 but are able to make code changes to your TLS 1.2 implementation, I suppose after this PR is fine, though I'm confused why you don't use your code change powers to get TLS 1.3.

Case 3 also isn't helped by this draft, before or after the PR. The existing TLS 1.2 PSK endpoints don't know how to do the IPSK derivation, so, one way or another, the dual-1.2/1.3 endpoints need to speak the old-style derivation at TLS 1.2, not the new style. Depending on how exactly the PSKs are configured, one strategy is to simply configure two separate PSKs, one legacy for TLS 1.2 and one new for TLS 1.3+. If that's not an option and you truly only get one root secret, you are sad. You could pass the same root secret into the legacy and modern path, at the cost of some cross-protocol risk. That's the problem I was trying to address [here](https://tools.ietf.org/html/draft-davidben-tls-universal-psk-00#section-3).

Case 4 is a harder version of case 3. The library's options are either to say, as a compatibility hack, disable TLS 1.3 for applications that called the `SetTLS12PreSharedKey()` function, or you've got the "one root secret" problems above.

Author: I agree with your points above. It seems to me that the most realistic case is (1), and, at that point, you might as well just prohibit importing for TLS 1.2 and prior versions. That lets us drop a lot of cruft. (I applied that change in the latest update!)
